client/serviceregistration/checks/client.go

```diff
@@ -163,7 +163,14 @@ func (c *checker) checkHTTP(ctx context.Context, qc *QueryContext, q *Query) *st
 		qr.Status = structs.CheckFailure
 		return qr
 	}
-	request.Header = q.Headers
+	for header, values := range q.Headers {
+		for _, value := range values {
+			request.Header.Add(header, value)
+		}
+	}
+
+	request.Host = request.Header.Get("Host")
```

Comment:
Note for future readers of this PR: this is safe without checking for it being empty because:

> 	// For client requests, Host optionally overrides the Host
>	// header to send. If empty, the Request.Write method uses
>	// the value of URL.Host. Host may contain an international
>	// domain name.